File: leetcode/easy/141-Linked-List-Cycle.py
#!/usr/bin/python3
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: list) -> bool:
        if head == None or head.next == None:
            return False
        slow = head.next
        fast = head.next.next
        while fast != None or slow != None:
            if fast == slow:
                return True
            if slow.next == None or fast.next == None or fast.next.next == None:
                return False
            else:
                slow = slow.next
                fast = fast.next.next
        return False

File: neetcode150/easy/141-Linked-List-Cycle.cpp
#include <algorithm>

/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};
 
 /**
  * \c Solution
  * 
  * 1. Use fase and slow pointer
  * 2. while fast and fast->next so that you can call fast->next->next inside the loop
  * 
  * Time  : O(n)
  * Space : O(1)
  */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if (head == nullptr) {
            return false;
        }

        ListNode* slow = head;
        ListNode* fast = head->next;
        while (fast != nullptr && fast->next != nullptr) {
            if (slow == fast) {
                return true;
            }
            slow = slow->next;
            fast = fast->next->next;
        }
        return false;
    }
};